app/buck2_http/src/client.rs

```diff
@@ -134,11 +139,14 @@ impl HttpClient {
                 HttpError::SendRequest { uri, source: e }
             }
         })?;
-        Ok(
-            resp.map(|body| {
-                CountingStream::new(body, self.stats.downloaded_bytes().dupe()).boxed()
-            }),
-        )
+        Ok(resp.map(move |body| {
+            CountingStream::new(body, self.stats.downloaded_bytes().dupe())
+                .inspect(move |_| {
+                    // Ensure we keep a concurrent request permit alive until the stream is consumed
+                    let _guard = &semaphore_guard;
+                })
```

Comment:
This is a little cheeky and might be clearer as a stream transformer struct, but that would probably take about 4x as much code. Let me know if that'd be preferred.

Reply:
These requests are typically downloading potentially large files. We don't want to release semaphore credit until the response has been *completed*, which is after the stream has been consumed in full and dropped. Without stream ownership of the guard, it would be released as soon as the response *begins*, and we could still end up with arbitrarily large numbers of concurrent [attempted] requests, which is what this PR seeks to avoid.